matchzoo/models/esim.py

```diff
@@ -0,0 +1,218 @@
+        cls_input = keras.layers.Dropout(
+            rate=self._params['dropout_rate'])(cls_input)
+        # Output layer
+        mlp = self._make_multi_layer_perceptron_layer()(cls_input)
+        mlp = keras.layers.Dropout(
+            rate=self._params['dropout_rate'])(mlp)
+        inputs = [input_left, input_right]
+        # mlp = keras.layers.Lambda(lambda x: K.tf.Print(x, [x]))(mlp)
+        x_out = self._make_output_layer()(mlp)
+        self._backend = keras.Model(inputs=inputs, outputs=x_out)
+
+    def _attention(self, encoded_left,
+                   left_mask,
+                   encoded_right,
+                   right_mask):
+        # Interaction
+        embed_cross = keras.layers.Dot(axes=2)([encoded_left, encoded_right])
+        cross_shape = [self._params['input_shapes'][0][0], self._params['input_shapes'][1][0]]
+        embed_cross = keras.layers.Reshape(cross_shape)(embed_cross)
+        left2right_attn = self._masked_softmax(embed_cross, left_mask,
+                                               self._params['input_shapes'][0][0],
+                                               self._params['input_shapes'][1][0])
+        # left2right_attn = keras.layers.Lambda(lambda x: K.tf.Print(x, [x]))(left2right_attn)
+
+        right2left_attn = self._masked_softmax(keras.layers.Permute([2, 1])(embed_cross), right_mask,
+                                               self._params['input_shapes'][1][0],
+                                               self._params['input_shapes'][0][0])
+
+        attended_left = self._weighted_sum(left2right_attn,
+                                           encoded_right,
+                                           left_mask)
+        attended_right = self._weighted_sum(right2left_attn,
+                                            encoded_left,
+                                            right_mask)
+        return attended_left, attended_right
+
+    def _masked_softmax(self, input, mask, att_len, base_len):
+        inf = 1e8
+
+        def tile_mask(x):
+            res = inf * K.tile(K.expand_dims(1 - x, axis=1), [1, base_len, 1])
+            return res
+
+        # def fill_mask(x):
+        #     return inf * x
+
+        tiled_mask = keras.layers.Lambda(tile_mask)(mask)
+        flattened_mask = keras.layers.Flatten()(tiled_mask)
+        flattened_input = keras.layers.Flatten()(input)
+        softmax_res = keras.layers.Softmax() \
+            (keras.layers.Subtract()([flattened_input, flattened_mask]))
+        output = keras.layers.Reshape([att_len, base_len])(softmax_res)
+        return output
+
+    def _weighted_sum(self, weights, tensor, mask):
+        weighted_sum = keras.layers.Dot(axes=[2, 1])([weights, tensor])
+        expanded_mask = keras.layers.Lambda(expand_tensor)(mask)
+        return keras.layers.Multiply()([expanded_mask, weighted_sum])
+
+
+def get_mask(x):
+    boolean_mask = K.not_equal(x, 0)
+    return K.cast(boolean_mask, K.dtype(x))

```

Comment:
let others reuse this piece of code

Reply:
These piece of code are very brief. Should I package the each function as a layer like MatchingLayer?

Reply:
I think keeping this local is perfectly fine. We could refactor it later when there are more uses.